### PR TITLE
feat: implicit allowlist for co-located test files in scope check

### LINKED ISSUE #105: Scope check: implicitly include co-located test files when their source file is in scope
## Context

The `agent-safe-scope.yml` check enforces that all files touched by an agent-safe PR appear in the issue body's `## Files to touch` section. This is the load-bearing mechanical gate distinguishing autonomous-merge PRs from human-review PRs.

Today (2026-04-26), PR #104 implementing #35 (UI Dashboard removal) FAILed scope check because it touched `ui/src/api.test.js` while only `ui/src/api.js` was listed. The file change was legitimate — `api.test.js` had test fixtures depending on the exports being deleted from `api.js`; re-anchoring the fixtures was the only way to preserve test coverage without leaving dead references.

This is a structural pattern, not a one-off:

- Deleting an exported function/class from `module.X` forces test changes in any co-located test file (`module.test.X`, or `tests/unit/test_module.py`) that referenced those exports.
- The pattern recurs on every refactor, deletion, or rename involving a tested module.
- The current rule treats co-located test changes as out-of-scope, forcing either issue-body amendments (workaround we used today) or human review (defeats the autonomous-merge premise for routine refactor work).

Same pattern was anticipated under #77's broader scope-rule maturation work but was not specifically addressed by #77, #83, or #90.

## Design questions to resolve

These should be locked during a `design-review` pass before implementation begins:

### 1. Scope of the relaxation

Which co-location conventions does the rule recognize?

- **JS/TS co-located:** `<file>.test.{js,jsx,ts,tsx}` next to `<file>.{js,jsx,ts,tsx}` (Vitest convention used in `ui/`).
- **Python sibling-directory:** `tests/unit/test_<file>.py` for `src/starter/<file>.py` (pytest convention used in this repo's backend).
- **Python co-located:** `<file>_test.py` next to `<file>.py` (Go-style convention; not used in this repo currently but worth knowing).

Recommended: cover JS/TS co-located AND Python sibling-directory, since both conventions are in active use here. Skip co-located Python (not used). Decision needs locking before implementation.

### 2. Implicit allowlist or explicit syntax

Two implementation shapes:

- **Implicit:** if `path/to/foo.js` is in `## Files to touch`, the parser also implicitly accepts `path/to/foo.test.js` (and Python equivalents). No issue-body change needed; rule infers.
  - Pro: zero burden on issue authors; matches mental model of "co-located test follows source."
  - Con: parser needs to recognize test-file naming patterns per language; more code in `check_agent_safe_scope.py`.
- **Explicit:** issue authors continue to list both files when both are touched. Tooling/docs nudge them to remember.
  - Pro: simpler parser; explicit is honest.
  - Con: continues today's failure mode (issue authors won't remember; PRs will continue to fail scope check; back to today's amendment workaround).

Recommended: implicit. The whole point of the autonomous path is that the issue-worker doesn't need to anticipate every co-located cleanup. Decision needs locking.

### 3. Edge cases for the implicit form

If implicit allowlist is chosen:

- What if the test file is in scope but the source file isn't? (Probably should still PASS — testing-only changes are legitimate, e.g., adding test coverage.)
- What if a file matching the test naming pattern exists but isn't actually a test (e.g., someone names a regular file `foo.test.js` for unrelated reasons)? (Edge case; probably accept the false positive — the cost is minor; the alternative is parsing the file content.)
- Snapshot files (`__snapshots__/<file>.test.X.snap`) — also implicitly in scope? (Recommended: yes, same reasoning.)

## Files to touch

- `scripts/check_agent_safe_scope.py` — extend the scope-matching logic to recognize co-located test files per the design pass's chosen approach.
- `tests/unit/test_check_agent_safe_scope.py` — add test cases covering each co-location convention adopted, plus the edge cases from design question 3.
- `CLAUDE.md` — if the existing scope-check documentation enumerates the rule, update it to mention the implicit-test-file behavior.
- `.claude/agents/issue-worker.md` — if the agent definition references the scope rule's behavior in its workflow, update for accuracy.

## Acceptance criteria

- [ ] Design pass locks: (a) which co-location conventions are recognized; (b) implicit allowlist vs explicit syntax; (c) edge case handling.
- [ ] `scripts/check_agent_safe_scope.py` matches the locked design.
- [ ] Test coverage in `tests/unit/test_check_agent_safe_scope.py` exercises every recognized co-location convention plus negative cases (e.g., `foo.test.js` is NOT implicitly accepted when `foo.js` isn't in scope).
- [ ] PR #104's exact failure mode (touching `ui/src/api.test.js` when `ui/src/api.js` is in scope) would have PASSed under the new rule. Validate via a unit test recreating the case.
- [ ] CLAUDE.md and (if applicable) issue-worker.md documentation reflect the new behavior.
- [ ] `inv pre-push` passes; no regressions in existing scope-check tests.
- [ ] Verify against the next agent-safe PR with co-located test changes — scope check passes without issue-body amendment.

## Notes

- NOT `agent-safe` — modifies the scope-check tooling itself; needs human review on the recursive-applicability boundary.
- Workaround in use until this lands: amend issue body to add the co-located test file (see PR #104's resolution path).
- Spawned from PR #104's scope FAIL on 2026-04-26.
- Same family of work as #77 (scope-rule maturation), #83 (taxonomy + WARN levels), #90 (parser edge cases). Different layer — those addressed scope-rule semantics; this addresses an implicit-allowlist extension.
- Design pass first: this issue is filed `status:design-needed` deliberately. The three open design questions need explicit resolution before implementation.
- Soft preference: land before any major UI or backend refactor is queued. Each refactor that recurs the today-pattern is friction.

### PR DESCRIPTION
**Pull request by warlordofmars**
Closes #105

## Summary

Extends `scripts/check_agent_safe_scope.py` so that when a source file is listed in `## Files to touch`, its co-located test file is implicitly accepted. Removes the friction that produced PR #104's scope FAIL on 2026-04-26 — issue authors no longer need to enumerate both halves of a source/test pair.

## Approach

Two distinct rule shapes (locked in [#105's design comment](https://github.com/warlordofmars/agentcore-starter/issues/105#issuecomment-4323559512), implemented with two parallel match paths rather than a single parameterized rule):

- **Rule A — JS/TS same-directory `.test` infix.** `<dir>/<name>.<ext>` (ext ∈ {js,jsx,ts,tsx}) listed in scope implicitly accepts `<dir>/<name>.test.<ext>` and `<dir>/__snapshots__/<name>.test.<ext>.snap`. Same-directory only — does NOT cross directories.
- **Rule B — Python basename-keyed, fixed `tests/unit/` root.** Any `<basename>.py` in scope implicitly accepts `tests/unit/test_<basename>.py` (flat) and `tests/unit/**/test_<basename>.py` (nested). Cross-directory by design — pytest convention puts tests in a fixed root regardless of source location, covering both `src/starter/**` and `scripts/**` without enumerating either.

The implicit allowlist is **forward-direction only**. Listing source implies test, but listing test does NOT imply source. Production code changes always require explicit listing — preserves the property that the issue body documents the real production-code surface of change.

Implementation is purely additive: a new `_implicit_test_paths` helper computes the derived set per source entry, and `check_scope` runs an `_expand_implicit_tests` pass over `allowed_globs` before the existing match logic. Existing scope sources (`## Files to touch` parsing, area-label fallback, universal-allowed `CHANGELOG.md`) all behave identically.

## Test coverage

10 new unit tests in `tests/unit/test_check_agent_safe_scope.py`:

- 4 happy paths — Rule A (JS, TSX) and Rule B (flat, nested).
- 3 negative cases — forward-direction enforcement (JS + Python), cross-directory rejection.
- 3 helper-function tests — direct surface of `_implicit_test_paths` for JS / Python / non-source paths.

The first happy-path test (`test_implicit_test_rule_a_js_co_located_test_passes`) is the **PR #104 regression test** required by AC4 — `ui/src/api.js` listed alongside `ui/src/api.test.js` in the diff passes scope check without amending the issue body.

Snapshot test deferred per design — no `__snapshots__/` directories exist in the repo today (verified via `find ui -name __snapshots__ -type d`); the rule accommodates them prospectively.

## Doc updates

- **CLAUDE.md** — added a paragraph under the `agent-safe` label documentation describing both rule shapes and the forward-direction-only property.
- **`.claude/agents/issue-worker.md`** — verified by grep; the only references to "Files to touch" are for skill-selection path prediction (line 78) and the W1 push allowlist (lines 342–346), neither of which describes the scope-check rule's behaviour. No edits made; the file is untouched.

## Recursive applicability

This PR modifies the scope-check tooling itself; the new rules take effect post-merge for future PRs. The PR's own scope check runs under existing rules. Three files touched (`scripts/check_agent_safe_scope.py`, `tests/unit/test_check_agent_safe_scope.py`, `CLAUDE.md`) — all explicitly listed in #105's `## Files to touch`. Verified offline via `python scripts/check_agent_safe_scope.py --issue-body-file ... --diff-files-file ...` → PASS.

## Notes

NOT `agent-safe` — modifies the scope-check tooling itself; needs human review on the recursive-applicability boundary, per #105's note. Auto-merge intentionally NOT armed.

### CODE REVIEW

File: scripts/check_agent_safe_scope.py [hunk 1]
```diff
@@ -284,12 +284,114 @@ def _matches_any(path: str, globs: list[str]) -> bool:
...
+#
+# Forward-direction only: listing source implicitly accepts test, but
+# listing test does NOT implicitly accept source. Production code changes
+# always require explicit listing — preserves the property that the issue
+# body documents the real production-code surface of change.
+
+_JS_TS_EXTS = ("js", "jsx", "ts", "tsx")
+# Source filename pattern for Rule A. Must NOT already be a test file —
+# `<name>.<ext>` where `<name>` does not end with `.test`.
+_JS_TS_SOURCE_RE = re.compile(r"^(?P<name>.+?)\.(?P<ext>js|jsx|ts|tsx)$")
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply scripts/check_agent_safe_scope.py]: Fixed in 8f88a47 —  is now used to build  via an -d alternation, making the tuple the single source of truth for supported JS/TS extensions.

File: scripts/check_agent_safe_scope.py [hunk 2]
```diff
@@ -284,12 +284,114 @@ def _matches_any(path: str, globs: list[str]) -> bool:
...
+# Python source filename pattern for Rule B. Must NOT already be a test
+# file — `<basename>.py` where `<basename>` does not start with `test_`.
+_PY_SOURCE_RE = re.compile(r"^(?P<basename>[^/]+)\.py$")
+
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply scripts/check_agent_safe_scope.py]: Fixed in 8f88a47 — dropped  entirely;  is the only check needed for Rule B and is already used inline.

File: scripts/check_agent_safe_scope.py [hunk 3]
```diff
@@ -284,12 +284,114 @@ def _matches_any(path: str, globs: list[str]) -> bool:
...
+        basename = path.rpartition("/")[2].removesuffix(".py")
+        # Skip files that are already test files (basename starts with
+        # `test_`). Forward-direction only.
+        if basename and not basename.startswith("test_"):
+            # Flat: tests/unit/test_<basename>.py
+            derived.append(f"tests/unit/test_{basename}.py")
+            # Nested: any path under tests/unit/ ending in
+            # test_<basename>.py — modelled with a glob.
+            derived.append(f"tests/unit/**/test_{basename}.py")
+
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply scripts/check_agent_safe_scope.py:390]: Fixed in 8f88a47 — `_implicit_test_paths()` now early-returns when the entry contains glob metacharacters (`*`, `?`, `[`). Two tests added: `test_implicit_test_helper_skips_glob_entries` (direct) and `test_implicit_test_glob_in_files_to_touch_does_not_widen_scope` (end-to-end with `src/starter/api/*.py` rejecting an unrelated test).

File: CLAUDE.md [hunk 4]
```diff
@@ -588,6 +588,27 @@ This section defines the taxonomy.
...
+  - **Python** — listing any `<basename>.py` (under `src/starter/**`,
+    `scripts/**`, or anywhere) implicitly accepts
+    `tests/unit/test_<basename>.py` and any nested
+    `tests/unit/**/test_<basename>.py` (matches the repo's pytest
+    convention of a fixed test root regardless of source location).
+
+  The implicit allowlist is **forward-direction only**: listing the
+  source implicitly accepts the test, but listing the test does NOT
+  implicitly accept the source. Production code changes always require
+  explicit listing — this preserves the property that the issue body
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply CLAUDE.md]: Fixed in 8f88a47 — CLAUDE.md now notes that Python entries with `test_`-prefixed basenames are skipped (forward-direction only) and that glob entries don't derive implicit tests.

File: scripts/check_agent_safe_scope.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (warlordofmars) [reply scripts/check_agent_safe_scope.py]: (Reply formatting note — backticks in the previous reply got eaten.) Fixed in 8f88a47 — `_JS_TS_EXTS` is now used to build `_JS_TS_SOURCE_RE` via an `re.escape`-d alternation, making the tuple the single source of truth for supported JS/TS extensions.

File: scripts/check_agent_safe_scope.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (warlordofmars) [reply scripts/check_agent_safe_scope.py]: (Reply formatting note — backticks in the previous reply got eaten.) Fixed in 8f88a47 — dropped `_PY_SOURCE_RE` entirely; `path.endswith('.py')` is the only check needed for Rule B and is already used inline.

File: scripts/check_agent_safe_scope.py [hunk 5]
```diff
@@ -284,12 +284,131 @@ def _matches_any(path: str, globs: list[str]) -> bool:
...
+      - the path already looks like a test file (e.g. `foo.test.js` or
+        `tests/unit/test_foo.py` — forward-direction only),
+      - the entry contains glob metacharacters (`*`, `?`, `[`) — globs are
+        not concrete paths and would derive an over-broad implicit set
+        (e.g. `src/starter/**` would map to `tests/unit/test_**.py`).
+    """
+    derived: list[str] = []
+
+    # Strip leading "./" if present; otherwise the path stays as-is.
+    path = source_path.removeprefix("./")
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply scripts/check_agent_safe_scope.py]: Fixed in 98f39ee — dropped the `./`-prefix strip from `_implicit_test_paths()`. You're right: `check_scope()` / `_matches_any()` don't normalise paths either way, so the strip created asymmetric semantics where an entry like `./ui/src/api.js` would be rejected as out-of-scope by the matcher but still derive a passing implicit test. Keeping one consistent semantics is the right call.

File: scripts/check_agent_safe_scope.py [hunk 6]
```diff
@@ -284,12 +284,131 @@ def _matches_any(path: str, globs: list[str]) -> bool:
...
+    accepted alongside it.
+
+    Returns an empty list when:
+      - the source path is not a recognised source file (e.g. `.md`,
+        `.yml`),
+      - the path already looks like a test file (e.g. `foo.test.js` or
+        `tests/unit/test_foo.py` — forward-direction only),
+      - the entry contains glob metacharacters (`*`, `?`, `[`) — globs are
+        not concrete paths and would derive an over-broad implicit set
+        (e.g. `src/starter/**` would map to `tests/unit/test_**.py`).
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply scripts/check_agent_safe_scope.py]: Fixed in 98f39ee — docstring example now uses `src/starter/**/*.py`, which is the realistic shape that would hit Rule B's branch and trigger the guard. The previous `src/starter/**` example would never have entered Rule B (no `.py` extension), so the guard's documented purpose was harder to map onto a concrete failure mode.
